src/with_serde.rs

```diff
@@ -0,0 +1,184 @@
+use crate::HashBag;
+use core::fmt;
+use core::hash::{BuildHasher, Hash};
+use core::marker::PhantomData;
+use serde::de::{Error, MapAccess, SeqAccess, Visitor};
+use serde::ser::{SerializeSeq, SerializeStruct, Serializer};
+use serde::Deserializer;
+use serde::{Deserialize, Serialize};
+
+pub(crate) struct HashBagVisitor<T, S> {
+    marker: PhantomData<fn() -> HashBag<T, S>>,
+}
+
+impl<T, S> HashBagVisitor<T, S>
+where
+    T: Eq + Hash,
+    S: BuildHasher + Clone,
+{
+    fn new() -> Self {
+        HashBagVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'de, T, S> Visitor<'de> for HashBagVisitor<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    type Value = HashBag<T, S>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a HashBag")
+    }
+
+    fn visit_seq<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: SeqAccess<'de>,
+    {
+        let mut bag: HashBag<T, S> =
+            HashBag::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
+
+        while let Some(entry) = access.next_element::<EntryWrapper<T>>()? {
+            bag.insert_many(entry.entry, entry.count);
+        }
+
+        Ok(bag)
+    }
+}
+
+impl<'de, T, S> Deserialize<'de> for HashBag<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(HashBagVisitor::<T, S>::new())
```

Comment:
Hmm, wouldn't this make more sense using `deserialize_map`, and having `T` be the key and `count` be the value? That'd also save you the wrapper type.

Reply:
Keys to maps need to be string-like (at least in JSON), although serde will not complain about this at compile time. I had the same idea and ran into issues when doing some testing with JSON. That is to say, the following struct would produce the following invalid JSON.

```rust
struct MyStruct {
  foo: String,
}
```

```json
{
  {"foo": "bar"} : 42,
  {"foo": "baz"} : 84,
}
```

Reply:
Oh, interesting. But at the same time, surely `HashMap` has the same problem if you try to use a key that isn't legal in key position for JSON?

Reply:
Ya, after some testing, it would seem that a regular hashmap with keys that aren't strings (or at least that aren't json primitives) can't be jsonified. It also doesn't seem to help if the type implements `ToString` and/or `From<String>`/`From<&str>`.

I would imagine (with no way of verifying) that serde is intended to map Rust's map as closely to JSON's maps as possible. However, the fact that `HashBag` uses a map seems to be of little consequence to the user. In fact, `HashBag` could be implemented using a `HashSet<Entry<T>>` where `Entry` encapsulates an instance of `T` and its count and uses its held instance of `T` for `Hash` and `Eq`.